File: hdf-converters/converters/oscal-to-hdf/typescript/convert.ts
import { convertOscalCatalogToHdf } from './converter-catalog.js';
import { convertOscalComponentToHdf } from './converter-component.js';
import { convertOscalPoamToHdf } from './converter-poam.js';
import { convertOscalSapToHdf } from './converter-sap.js';
import { convertOscalSarToHdf } from './converter-sar.js';
import { convertOscalSspToHdf } from './converter-ssp.js';
import { detectOscalDocumentType } from './detect.js';

/**
 * Domain-level auto-detect entry point: detects the OSCAL document type and
 * dispatches to the matching per-type converter, returning the serialized HDF
 * document (baseline / system / plan / results / amendments). Mirrors the Go
 * ConvertOSCALToHDF seam shared by the CLI auto-detect wrapper and the snapshot
 * harness.
 *
 * Profile resolution requires a separate catalog input, so it cannot be handled
 * from a single document; callers that need it must invoke convertOscalProfileToHdf
 * directly with both inputs.
 */
export async function convertOscalToHdf(input: string): Promise<string> {
  const docType = detectOscalDocumentType(input);
  switch (docType) {
    case 'catalog':
      return convertOscalCatalogToHdf(input);
    case 'component-definition':
      return convertOscalComponentToHdf(input);
    case 'system-security-plan':
      return convertOscalSspToHdf(input);
    case 'assessment-plan':
      return convertOscalSapToHdf(input);
    case 'assessment-results':
      return convertOscalSarToHdf(input);
    case 'plan-of-action-and-milestones':
      return convertOscalPoamToHdf(input);
    case 'profile':
      throw new Error(
        'oscal profile requires a separate catalog input; call convertOscalProfileToHdf with both documents',
      );
    default:
      throw new Error(`unsupported OSCAL document type ${String(docType)}`);
  }
}

File: hdf-converters/converters/oscal-to-hdf/go/convert.go
package oscal

import "fmt"

// ConvertOSCALToHDF is the domain-level auto-detect entry point: it detects the
// OSCAL document type and dispatches to the matching per-type converter,
// returning the appropriate HDF document (baseline / system / plan / results /
// amendments) as interface{}. This is the single seam the CLI auto-detect
// wrapper and the snapshot harness share.
//
// Profile resolution requires a separate catalog input, so it cannot be handled
// from a single document; callers that need it must invoke ConvertProfileToHDF
// directly with both inputs.
func ConvertOSCALToHDF(input []byte, converterVersion string) (interface{}, error) {
	docType, err := DetectDocumentType(input)
	if err != nil {
		return nil, err
	}

	switch docType {
	case "catalog":
		return ConvertCatalogToHDF(input, converterVersion)
	case "component-definition":
		return ConvertComponentDefinitionToHDF(input, converterVersion)
	case "system-security-plan":
		return ConvertSSPToHDF(input, converterVersion)
	case "assessment-plan":
		return ConvertAssessmentPlanToHDF(input, converterVersion)
	case "assessment-results":
		return ConvertAssessmentResultsToHDF(input, converterVersion)
	case "plan-of-action-and-milestones":
		return ConvertPOAMToHDF(input, converterVersion)
	case "profile":
		return nil, fmt.Errorf("oscal profile requires a separate catalog input; call ConvertProfileToHDF with both documents")
	default:
		return nil, fmt.Errorf("unsupported OSCAL document type %q", docType)
	}
}
